### PR TITLE
Automate generating checksum files for release

### PR DESCRIPTION
**Pull request by mitalipaygude**
*Issue #, if available:*
#1656 
*Description of changes:*

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: scripts/generate_checksum.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+SHASUM_FILES="SHASUMS"
+FILE_PREFIX="eksctl-anywhere"
+
+DARWIN_TAR="darwin-amd64.tar.gz"
+LINUX_TAR="linux-amd64.tar.gz"
+
+release_manifest=$(curl --silent 'https://anywhere-assets.eks.amazonaws.com/releases/eks-a/manifest.yaml')
+
+latest_release_version=$(echo $release_manifest | yq e '.spec.latestVersion' - -v)
+echo $latest_release_version
```
Comment (chrisdoherty4) [inline scripts/generate_checksum.sh:25]: I don't know where the script is run but its typically good practice to give context to echo statements like this unless they're used in some bash pipe cmd (doesn't look like it judging by the rest of the file).

```suggestion
echo "Latest release version: $latest_release_version"
```

Same applies to some of the other naked echo statements.

File: scripts/generate_checksum.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+cd $SHASUM_FILES
+
+SHA256_FILENAME="{FILE_PREFIX}-{latest_release_version}-sha256sums.txt"
+SHA512_FILENAME="{FILE_PREFIX}-{latest_release_version}-sha512sums.txt"
+
+echo "${linux_sha256} ${FILE_PREFIX}-${latest_release_version}-${LINUX_TAR}" >> SHA256_FILENAME
+echo "${darwin_sha256} ${FILE_PREFIX}-${latest_release_version}-${DARWIN_TAR}" >> SHA256_FILENAME
+
+echo "${linux_sha512} ${FILE_PREFIX}-${latest_release_version}-${LINUX_TAR}" >> SHA512_FILENAME
+echo "${darwin_sha512} ${FILE_PREFIX}-${latest_release_version}-${DARWIN_TAR}" >> SHA512_FILENAME
```
Comment (chrisdoherty4) [inline scripts/generate_checksum.sh:47]: Is everything after `>>` meant to be a variable? I know this is a draft so you're probably working out kinks.

File: scripts/generate_checksum.sh (hunk 1 above)
Comment (mitalipaygude) [reply scripts/generate_checksum.sh]: Hey this was for my debugging, wont be a part of the final script.

File: scripts/generate_checksum.sh (hunk 2 above)
Comment (mitalipaygude) [reply scripts/generate_checksum.sh]: yeah I missed the `$` sign here, its updated now

File: scripts/generate_checksum.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+SHASUM_FILES="shasums-eksctl-anywhere"
+FILE_PREFIX="eksctl-anywhere"
+
+DARWIN_TAR="darwin-amd64.tar.gz"
+LINUX_TAR="linux-amd64.tar.gz"
+
+mkdir -p $SHASUM_FILES
+cd $SHASUM_FILES
+
+release_manifest=$(curl --silent 'https://anywhere-assets.eks.amazonaws.com/releases/eks-a/manifest.yaml' -o release.yaml)
```
Comment (abhay-krishna) [inline scripts/generate_checksum.sh:25]: Why are we assigning this to a variable?

File: scripts/generate_checksum.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+latest_release_version=$(cat release.yaml | yq e '.spec.latestVersion')
+
+latest_release=$(cat release.yaml | yq e '.spec.releases[] | select(.version == "'$latest_release_version'")' >> latest_release.yaml)
+
+linux_sha256=$(cat latest_release.yaml | yq e '.eksABinary.linux.sha256')
```
Comment (abhay-krishna) [inline scripts/generate_checksum.sh:31]: There must be a way to put this in a loop over [256, 512] since we're doing very similar operations.

File: scripts/generate_checksum.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+DARWIN_TAR="darwin-amd64.tar.gz"
+LINUX_TAR="linux-amd64.tar.gz"
+
+mkdir -p $SHASUM_FILES
+cd $SHASUM_FILES
+
+release_manifest=$(curl --silent 'https://anywhere-assets.eks.amazonaws.com/releases/eks-a/manifest.yaml' -o release.yaml)
+
+latest_release_version=$(cat release.yaml | yq e '.spec.latestVersion')
```
Comment (abhay-krishna) [inline scripts/generate_checksum.sh:27]: If piping yaml to `yq`, we need to put a dash after the yq expression to make it read inline string.
```suggestion
latest_release_version=$(cat release.yaml | yq e '.spec.latestVersion' -)
```

Comment (mitalipaygude) [reply scripts/generate_checksum.sh:27]: I think that changed since 4.18.1
https://mikefarah.gitbook.io/yq/

Comment (abhay-krishna) [reply scripts/generate_checksum.sh:27]: Ah I see, we need to make sure we have a check for that then or enforce 4.18+. similar to [this](https://github.com/aws/eks-anywhere-build-tooling/blob/main/build/lib/generate_projects_list.sh#L9-L18) but not quite the same.

Also not everyone might have yq locally, so testing that edge case would be good too.

File: scripts/generate_checksum.sh [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+EXPECTED_YQ_MINOR_VERSION=18
+
+if [ $(($CURRENT_YQ_MAJOR_VERSION)) -lt $(($EXPECTED_YQ_MAJOR_VERSION)) ]; then
+    echo "ERROR: Current yq major version v$CURRENT_YQ_MAJOR_VERSION.x is older than expected v$EXPECTED_YQ_MAJOR_VERSION.x"
+    echo "Please install the latest version of yq from $YQ_LATEST_RELEASE_URL"
+    exit 1
+fi
+
+if [ $CURRENT_YQ_MAJOR_VERSION == $(($EXPECTED_YQ_MAJOR_VERSION)) ] && [ $CURRENT_YQ_MINOR_VERSION -lt $(($EXPECTED_YQ_MINOR_VERSION)) ]; then
+    echo "ERROR: Current yq minor version v$EXPECTED_YQ_MAJOR_VERSION.$CURRENT_YQ_MINOR_VERSION.x is older than expected v$EXPECTED_YQ_MAJOR_VERSION.$EXPECTED_YQ_MINOR_VERSION.x"
```
Comment (mrajashree) [inline scripts/generate_checksum.sh:50]:
```suggestion
    echo "ERROR: Current yq minor version v$CURRENT_YQ_MAJOR_VERSION.$CURRENT_YQ_MINOR_VERSION.x is older than expected v$EXPECTED_YQ_MAJOR_VERSION.$EXPECTED_YQ_MINOR_VERSION.x"
```

File: scripts/generate_checksum.sh [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+latest_release=$(cat release.yaml | yq e '.spec.releases[] | select(.version == "'$latest_release_version'")' >> latest_release.yaml)
+
+for sha in 256 512
+do
+
+  linux_sha=$(cat latest_release.yaml | yq e ".eksABinary.linux.sha${sha}")
+  darwin_sha=$(cat latest_release.yaml | yq e ".eksABinary.darwin.sha${sha}")
+
+  echo "${darwin_sha}  ${FILE_PREFIX}-${latest_release_version}-${DARWIN_TAR}" >> "${FILE_PREFIX}-${latest_release_version}-sha${sha}sums.txt"
```
[the remark this answers is not in the document]
Comment (mitalipaygude) [reply scripts/generate_checksum.sh:69]: we clean up these files at the very beginning of this script.

File: scripts/generate_checksum.sh [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+EXPECTED_YQ_MAJOR_VERSION=4
+EXPECTED_YQ_MINOR_VERSION=18
+
+if [ $(($CURRENT_YQ_MAJOR_VERSION)) -lt $(($EXPECTED_YQ_MAJOR_VERSION)) ]; then
+    echo "ERROR: Current yq major version v$CURRENT_YQ_MAJOR_VERSION.x is older than expected v$EXPECTED_YQ_MAJOR_VERSION.x"
+    echo "Please install the latest version of yq from $YQ_LATEST_RELEASE_URL"
+    exit 1
+fi
+
+if [ $CURRENT_YQ_MAJOR_VERSION == $(($EXPECTED_YQ_MAJOR_VERSION)) ] && [ $CURRENT_YQ_MINOR_VERSION -lt $(($EXPECTED_YQ_MINOR_VERSION)) ]; then
```
Comment (mrajashree) [inline scripts/generate_checksum.sh:49]: nit: since the check on line 43 already compares the major versions, I think we won't need to compare `if [ $CURRENT_YQ_MAJOR_VERSION == $(($EXPECTED_YQ_MAJOR_VERSION)) ]` again right?

File: scripts/generate_checksum.sh [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+mkdir -p $SHASUM_FILES
+cd $SHASUM_FILES
+curl --silent 'https://anywhere-assets.eks.amazonaws.com/releases/eks-a/manifest.yaml' -o release.yaml
+
+latest_release_version=$(cat release.yaml | yq e '.spec.latestVersion')
+
+latest_release=$(cat release.yaml | yq e '.spec.releases[] | select(.version == "'$latest_release_version'")' >> latest_release.yaml)
+
+for sha in 256 512
```
[the remark this answers is not in the document]
Comment (abhay-krishna) [reply scripts/generate_checksum.sh:63]: c12656e8736166eaead65e0579afd8e8

[the remark this answers is not in the document]
Comment (mitalipaygude) [reply scripts/generate_checksum.sh:63]: LOL

[the remark this answers is not in the document]
Comment (mitalipaygude) [reply scripts/generate_checksum.sh:63]: this bot is giving me the creeps

File: scripts/generate_checksum.sh (hunk 8 above)
Comment (mitalipaygude) [reply scripts/generate_checksum.sh]: we are just checking for less than above for the major versions, but say we have a 5.1 comparing to 4.18, the minor versions should be checked only when the major version is 4, and nothing else.
